Add a mesh generator to delete elements near a mesh

GiudGiud opened the pull request:
refs #32138

GiudGiud committed:
Add a mesh generator to delete elements near another mesh

GiudGiud committed:
Add node heuristic for deleting elements near a mesh

GiudGiud committed:
Add test for cutting a 2D mesh by another 2D mesh refs #32138

GiudGiud committed:
Add documentation
refs #32138

GiudGiud committed:
Add missing headers for unity build

lindsayad commented on framework/src/meshgenerators/DeleteElementsNearMeshGenerator.C:
```diff
@@ -0,0 +1,128 @@
+  InputParameters params = ElementDeletionGeneratorBase::validParams();
+
+  params.addClassDescription(
+      "Removes elements lying 'near' another mesh. The proximity is examined by the distance from "
```
```suggestion
      "Removes elements lying \"near\" another mesh. The proximity is examined by the distance from "
```

lindsayad commented on framework/doc/content/source/meshgenerators/DeleteElementsNearMeshGenerator.md:
```diff
@@ -0,0 +1,20 @@
+A side Gauss quadrature is created on all the element external sides of the `proximity_mesh`. The quadrature points from this
+quadrature are gathered into a K-nearest neighbor object.
+
+Then this KNN object provides the distance from the `proximity_mesh` to the element by computing the distance from
```
```suggestion
Then this KNN object provides the distance from the `proximity_mesh` to the element by computing the distance to
```

lindsayad commented on framework/src/meshgenerators/DeleteElementsNearMeshGenerator.C:
```diff
@@ -0,0 +1,128 @@
+  params.addParam<MooseEnum>(
+      "side_order",
+      options_order,
+      "Order of the face quadrature used to find the nearest face in the 'proximity_mesh'");
```
```suggestion
      "Order of the face quadrature used to find the nearest face in the 'proximity_mesh'. Default is CONSTANT");
```

lindsayad commented on framework/src/meshgenerators/DeleteElementsNearMeshGenerator.C:
```diff
@@ -0,0 +1,128 @@
+  std::vector<Point> all_side_qps;
+  const auto order = getParam<MooseEnum>("side_order");
+  all_side_qps.reserve(_proximity_mesh->n_elem() * _proximity_mesh->spatial_dimension() * 2 *
+                       order);
```
Why this heuristic? Maybe add a comment with the answer

lindsayad commented on framework/src/meshgenerators/DeleteElementsNearMeshGenerator.C:
```diff
@@ -0,0 +1,128 @@
+  const auto order = getParam<MooseEnum>("side_order");
+  all_side_qps.reserve(_proximity_mesh->n_elem() * _proximity_mesh->spatial_dimension() * 2 *
+                       order);
+  for (const auto elem : _proximity_mesh->element_ptr_range())
```
```suggestion
  for (const auto * const elem : _proximity_mesh->element_ptr_range())
```

lindsayad commented on framework/src/meshgenerators/DeleteElementsNearMeshGenerator.C:
```diff
@@ -0,0 +1,128 @@
+      const std::unique_ptr<const Elem> face = elem->build_side_ptr(side_i);
+      std::unique_ptr<libMesh::FEBase> fe(
+          libMesh::FEBase::build(dim, libMesh::FEType(elem->default_order())));
+      libMesh::QGauss qface(dim - 1, Moose::stringToEnum<Order>(order));
```
I'd do the `Moose::stringToEnum<Order>` outside the loop

GiudGiud replied:
done

GiudGiud committed:
Address Alex review

- more const
- more auto
- add more comments
- better grammar
- move an expensive operation out of the loop

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3483433/), step Clang format on 6607b13 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/32139/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/32139/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9812b68460ec7d62f728fe7a8de39d4f15923af5`
